lib/galaxy/webapps/galaxy/api/dataset_collections.py

```diff
@@ -81,6 +86,38 @@ def show(self, trans, instance_type, id, **kwds):
             view='element'
         )
 
+    @expose_api
+    def contents(self, trans, id, **kwds):
+        """
+        Show direct child contents of indicated dataset collection parent id
+        GET /api/dataset_collection/{id}/contents
+
+        Optional pagination parameters
+            limit:  integer
+            offset: integer
+        """
+
+        limit = kwds.get('limit', None)
+        offset = kwds.get('offset', None)
+        decoded_id = trans.app.security.decode_id(id)
```

Comment:
You're sure the way you're using this you have a DatasetCollection ID and not a HistoryDatasetCollectionAssocation ID? It is easy to make these two overlap for all IDs in simple tests - but they definitely won't in production. The other API methods in this class take in an association ID and do security checks on the associated object. They also do security checks on the associated object - I'm nervous about that not being here... I don't know if it is enough to just check the datasets.

Reply:
No - I guess looking at your example - the collection ID is different than the HDCA type ID. So this seems fine. Nice work. I'm still nervous about security but let me think about that.

Reply:
```dictify_element_reference``` doesn't do any checks - so I the other methods that use it are counting on the check of the HDCA for security. I think that means your API route gives access to dataset contents without any checks unless I'm missing something.

Reply:
@jmchilton I wondered about security as well. As I see it, only the top level HDCAs have a direct connection to the history and therefore the user/security layer unless I misunderstand. 

I could maybe write a secondary security query by using a recursive check up the parents of the selected layer to find a matching user history if you think that would be appropriate. It'll give me an excuse to exploit recursive CTEs in a query.

Reply:
@jmchilton I've included a standard security check from the existing code-base as well as a lookup that makes sure the indicated collection is actually a member of the referenced hdca.